### PR TITLE
JSON and YAML: allow deserializing abstract types

### PR DESCRIPTION
**Pull request by asterite**
This is something that wasn't possible right now, and I think just trying to deserialize every concrete subtype makes sense. In a way this is similar to how unions are deserialized.

### CODE REVIEW

**Comment by bcardiff** [comment]
A way to prioritize subtypes should be considered. Like an `attribute DeserializePriority` or similar.

**Comment by asterite** [comment]
@bcardiff Maybe, I don't know. If you have many possible types that an API allows I think there shouldn't be ambiguous parsing.

But... should it be part of this PR?

And also, if you need an order you can always override `self.new` and do custom logic there.

**Comment by straight-shoota** [comment]
@bcardiff That sounds like adding a lot of complexity and is IMO not strictly necessary.

File: src/json/serialization.cr [hunk 1]
```diff
@@ -115,6 +115,12 @@ module JSON
   #   @a : Int32?
   # end
   # ```
+  #
+  # ### Abstract types
+  #
+  # When deserializing an abstract type all concrete subtypes will be
```
Comment (RX14) [inline src/json/serialization.cr]: Perhaps say explicitly "when including this module in an abstract type"?

**Pull request closed without merging**